File: internal/domain/ranking.go
package domain

import (
	"math"
	"net/url"
	"sort"
	"strings"
	"time"
)

func ScoreItems(items []SourceItem, now time.Time) []ScoredItem {
	out := make([]ScoredItem, 0, len(items))
	for _, item := range items {
		age := ageHours(now, item.PublishedAt)
		freshness := math.Exp(-age / 24)
		rank := (item.Score + 0.5*item.Comments) * freshness
		out = append(out, ScoredItem{SourceItem: item, AgeHours: age, RankScore: rank})
	}
	return out
}

func ClusterByURL(items []ScoredItem) []Cluster {
	m := map[string]*Cluster{}
	order := 0
	for _, item := range items {
		key := canonicalizeURL(item.URL)
		if key == "" {
			key = string(item.Source) + ":" + item.ID
		}
		if c, ok := m[key]; ok {
			c.Items = append(c.Items, item)
			c.RankScore += item.RankScore
			if item.RankScore > c.Items[0].RankScore {
				c.Title = item.Title
			}
			continue
		}
		order++
		cl := &Cluster{
			ID:           "cluster_" + itoa(order),
			CanonicalURL: key,
			Title:        item.Title,
			Items:        []ScoredItem{item},
			RankScore:    item.RankScore,
		}
		m[key] = cl
	}
	out := make([]Cluster, 0, len(m))
	for _, c := range m {
		out = append(out, *c)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].RankScore > out[j].RankScore })
	return out
}

func canonicalizeURL(raw string) string {
	u, err := url.Parse(raw)
	if err != nil {
		return raw
	}
	u.Fragment = ""
	u.Path = strings.TrimSuffix(u.Path, "/")
	return u.String()
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	buf := [20]byte{}
	i := len(buf)
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	return string(buf[i:])
}

func ageHours(now time.Time, publishedAt string) float64 {
	pt, err := time.Parse(time.RFC3339, publishedAt)
	if err != nil {
		return 999
	}
	d := now.Sub(pt)
	if d < 0 {
		return 0
	}
	return d.Hours()
}

File: src/domain/ranking.ts
import { Cluster, ScoredItem, SourceItem } from "./model.js";
import { ageHours } from "../shared/time.js";

function canonicalizeUrl(url: string): string {
  try {
    const u = new URL(url);
    u.hash = "";
    if (u.pathname.endsWith("/")) u.pathname = u.pathname.slice(0, -1);
    return u.toString();
  } catch {
    return url;
  }
}

export function scoreItems(
  items: SourceItem[],
  prevState: Record<string, { score: number; comments: number }>,
  now: Date
): ScoredItem[] {
  return items.map((item) => {
    const key = `${item.source}:${item.id}`;
    const prev = prevState[key];
    const deltaScore = Math.max(0, item.score - (prev?.score ?? 0));
    const deltaComments = Math.max(0, item.comments - (prev?.comments ?? 0));
    const h = ageHours(now, item.publishedAt);
    const freshness = Math.exp(-h / 18);
    const trendScore =
      (Math.log1p(deltaScore) + 0.7 * Math.log1p(deltaComments) + 0.2 * Math.log1p(Math.max(0, item.score))) * freshness;

    return {
      ...item,
      ageHours: h,
      deltaScore,
      deltaComments,
      trendScore
    };
  });
}

export function clusterByUrl(scoredItems: ScoredItem[]): Cluster[] {
  const map = new Map<string, Cluster>();
  for (const item of scoredItems) {
    const url = canonicalizeUrl(item.url || item.rawUrl || `${item.source}:${item.id}`);
    const existing = map.get(url);
    if (existing) {
      existing.items.push(item);
      existing.deltaScore += item.deltaScore;
      existing.deltaComments += item.deltaComments;
      existing.trendScore += item.trendScore;
      if (item.trendScore > existing.items[0].trendScore) {
        existing.title = item.title;
      }
      continue;
    }

    map.set(url, {
      id: `cluster_${map.size + 1}`,
      canonicalUrl: url,
      title: item.title,
      items: [item],
      deltaScore: item.deltaScore,
      deltaComments: item.deltaComments,
      trendScore: item.trendScore
    });
  }

  return [...map.values()].sort((a, b) => b.trendScore - a.trendScore);
}

export function pickSections(clusters: Cluster[], maxTopics: number): {
  selected: Cluster[];
  topTopics: Cluster[];
  rising: Cluster[];
  deepDiscussion: Cluster[];
} {
  const selected = clusters.slice(0, maxTopics);
  const topTopics = [...selected].sort((a, b) => b.trendScore - a.trendScore).slice(0, 7);
  const rising = [...selected]
    .sort((a, b) => b.deltaScore - a.deltaScore || b.trendScore - a.trendScore)
    .slice(0, 5);
  const deepDiscussion = [...selected]
    .sort((a, b) => b.deltaComments - a.deltaComments || b.trendScore - a.trendScore)
    .slice(0, 5);

  return { selected, topTopics, rising, deepDiscussion };
}
